Fix crash with fully-qualified names in module-info.java import

msridhar opened the pull request:
Fixes #533 

We had no tests before for a `module-info.java` file, so we missed this regression.

msridhar commented:
I want to fix this in a slightly different place; will re-request review when it's done.

msridhar commented:
I also locally confirmed that NullAway 0.9.3 crashes on the caffeine master branch, but with this fix it compiles fine.

lazaroclapp commented on nullaway/src/main/java/com/uber/nullaway/NullabilityUtil.java:
```diff
@@ -307,7 +307,12 @@ public static TreePath findEnclosingMethodOrLambdaOrInitializer(TreePath path) {
+      // dereference.  E.g., in a `requires` clause in a module-info.java file, the fully-qualified
+      // package name is represented using `MemberSelectTree`s.  When the symbol is null, there is
+      // no possibility of a null dereference.
+      return false;
```
Is this the best place to filter these out? Or could we somehow check, before we call `mayBeNullFieldFromType(...)` if we truly have a field vs requires or any other `MemberSelectTree` using construct that isn't a method call or field dereference?

msridhar replied:
Yeah, I wish we could check better to enumerate the cases where we expect the symbol may be null.  `RequiresTree` was introduced in JDK 9, but with some slightly creative code and the `TreePath` to a `MemberSelectTree` we could probably detect that case.

That said, I do not see how it is possible to have a valid field dereference or method call `MemberSelectTree` that does not have a symbol.  If there is no symbol, that means the compiler could not resolve the field / method, and the code simply shouldn't compile, unless I am missing something.  We could add some stronger assertions around that.

My only hesitation is that adding stronger assertions might cause another regression / crash where things were working before.  But I guess that's better than not understanding what is going on 🙂 Let me tinker a bit.

msridhar commented:
@lazaroclapp please take another look.  Here's what I've done:

* Modified `mayBeNullExpr()` to check for a `null` symbol for the member-select case, and if it is `null`, conclude that the expression cannot be null.  We already made this assumption inside `matchMemberSelect()`, and I added more comments there explaining our reasoning.
* Changed `mayBeNullFieldFromType()` to only take a non-null `symbol` parameter and fail otherwise.

I would still rather not try to enumerate every single case where a `MemberSelectTree` can have a null symbol: I think that would be fragile, and I'm very confident that anytime the symbol is null it doesn't represent a real field dereference or method call.

lazaroclapp commented on nullaway/src/main/java/com/uber/nullaway/NullabilityUtil.java:
```diff
@@ -300,15 +300,15 @@ public static TreePath findEnclosingMethodOrLambdaOrInitializer(TreePath path) {
    * @param config NullAway config
    * @return true if based on the type, package, and name of the field, the analysis should assume
    *     the field might be null; false otherwise
+   * @throws NullPointerException if {@code symbol} is null
```
We really should be running NullAway on NullAway 😅 

(Brings to mind a certain saying in Spanish about a blacksmith having only wooden tools)

lazaroclapp reviewed (approved):
Makes sense to me. Still wish we could explicitly check and enumerate the member select cases, but that's hard while supporting JDK8 through 17 :)
